src/rust/engine/remote_provider/remote_provider_reapi/src/byte_store.rs

```diff
@@ -393,24 +395,48 @@ impl ByteStoreProvider for Provider {
         &self,
         digests: &mut (dyn Iterator<Item = Digest> + Send),
     ) -> Result<HashSet<Digest>, String> {
-        let request = remexec::FindMissingBlobsRequest {
-            instance_name: self.instance_name.as_ref().cloned().unwrap_or_default(),
-            blob_digests: digests.into_iter().map(|d| d.into()).collect::<Vec<_>>(),
-        };
+        let blob_digests = digests.into_iter().map(|d| d.into()).collect::<Vec<_>>();
+
+        let max_digests_per_request: usize = (4 * 1024 * 1024
+            - self
+                .instance_name
+                .as_ref()
+                .cloned()
+                .unwrap_or_default()
+                .len())
```

Comment:
Is the instance name encoded into the gRPC message as exactly its UTF-8 bytes? I'd personally be expecting an extra one or two bytes for either a varint-encoded length or a null terminator (the first I imagine). If we get unlucky with configuration, a request could exceed the 4MiB limit by 1 byte and that'd be very annoying!  (e.g. a 44 byte `instance_name` would make for a 4MiB request (`59918 * 70 + 44`) and thus any extra content would be too large)

I guess we could either:

- make this exactly precise (and a loop over different `instance_names` to the test)
- just add semi-arbitrary slop factor:

```suggestion
            - self
                .instance_name
                .as_ref()
                .cloned()
                .unwrap_or_default()
                .len() - 10)
```